src/server/ua_subscription_events.c

```diff
@@ -256,40 +256,55 @@ UA_Server_filterEvent(UA_Server *server, const UA_NodeId *eventNode, UA_EventFil
+            rpe.referenceTypeId = UA_NODEID_NUMERIC(0, UA_NS0ID_HASTYPEDEFINITION);
+            bpr = UA_Server_translateBrowsePathToNodeIds(server, &bp);
+        } else {
+            /* The list of BrowseNames is equivalent to a RelativePath that
+             * specifies forward references which are subtypes of the
+             * HierarchicalReferences ReferenceType. All Nodes followed by the
+             * browsePath shall be of the NodeClass Object or Variable. */
+
+            UA_STACKARRAY(UA_RelativePathElement, rpe, clause->browsePathSize);
+            memset(rpe, 0, sizeof(UA_RelativePathElement) * clause->browsePathSize);
+            for(size_t j = 0; j < clause->browsePathSize; j++) {
+                rpe[j].referenceTypeId = UA_NODEID_NUMERIC(0, UA_NS0ID_HIERARCHICALREFERENCES);
+                rpe[j].includeSubtypes = true;
+                rpe[j].targetName = clause->browsePath[j];
+            }
+            bp.relativePath.elements = rpe;
+            bp.relativePath.elementsSize = clause->browsePathSize;
+            bpr = UA_Server_translateBrowsePathToNodeIds(server, &bp);
         }
 
-        /* copy the value */
-        UA_Boolean whereClauseResult = UA_TRUE;
-        UA_Boolean whereClausesUsed = UA_FALSE;     /* placeholder until whereClauses are implemented */
-        UA_StatusCode retval = whereClausesApply(server, filter->whereClause,
-                                                 &notification->fields, &whereClauseResult);
-        if (retval == UA_STATUSCODE_BADNOTSUPPORTED)
-            whereClausesUsed = UA_TRUE;
-
-        if(whereClauseResult) {
-            retval = UA_Server_readValue(server, bpr.targets[0].targetId.nodeId,
-                                         &notification->fields.eventFields[i]);
-            if(retval != UA_STATUSCODE_GOOD)
-                UA_Variant_init(&notification->fields.eventFields[i]);
-
-            if(whereClausesUsed)
-                return UA_STATUSCODE_BADNOTSUPPORTED;
-        } else {
-            UA_Variant_init(&notification->fields.eventFields[i]);
-            /* TODO: better statuscode for failing at where clauses */
-            /* EventFilterResult currently isn't being used
-            notification->result.selectClauseResults[i] = UA_STATUSCODE_BADDATAUNAVAILABLE; */
+        if(bpr.statusCode != UA_STATUSCODE_GOOD || bpr.targetsSize == 0) {
+            UA_BrowsePathResult_deleteMembers(&bpr);
+            continue;
         }
+
+        UA_ReadValueId rvi;
+        UA_ReadValueId_init(&rvi);
+        rvi.nodeId = bpr.targets[0].targetId.nodeId;
+        rvi.attributeId = clause->attributeId;
+
+        UA_DataValue v = UA_Server_read(server, &rvi, UA_TIMESTAMPSTORETURN_NEITHER);
+        notification->fields.eventFields[i] = v.value;
         UA_BrowsePathResult_deleteMembers(&bpr);
+
+        UA_Boolean whereClauseResult = UA_TRUE;
+        return whereClausesApply(server, filter->whereClause, &notification->fields, &whereClauseResult);
```

Comment:
The `return` should be outside of the loop. Doing this ensures all data is copied correctly.